src/GoMath.ts

```diff
@@ -514,4 +514,168 @@ export class GoMath {
         });
         return GoMath.encodeMoves(moves);
     }
+
+    // This is intended to be an "easy to understand" method of generating a unique id
+    // for a board position.
+
+    // The id is the list of all the positions of the stones.
+
+    // There are in fact 8 possible ways to list the positions (all the rotations and
+    // reflections of the position).   The id is the lowest (alpha-numerically) of these.
+
+    // The "easy to understand" part is that the id can be compared visually to the
+    // board position
+
+    // The downside is that the id string can be moderately long for boards with lots of stones
+
+    public static positionId(
+        position: Array<Array<JGOFNumericPlayerColor>>,
+        height: number,
+        width: number,
+    ): string {
+        // The basic algorithm: list where the stones are, in a long string :)
+        let black_state = "";
+        let white_state = "";
+        for (let row = 0; row < height; row++) {
+            for (let col = 0; col < width; col++) {
+                if (position[row][col] === JGOFNumericPlayerColor.BLACK) {
+                    black_state += "B" + this.prettyCoords(row, col, height);
+                }
+                if (position[row][col] === JGOFNumericPlayerColor.WHITE) {
+                    white_state += "W" + this.prettyCoords(row, col, height);
+                }
+            }
+        }
+        let this_way = `${white_state}.${black_state}`;
+        let id = this_way;
+        // console.log("calc, id:", this_way, id);
+
+        // Take care of identical positions that are rotated/reflected: calculate
+        // using the same algorithm for each of the 8 transformations, and chose the
+        // alphanumerically shortest version as the actual id...
+
+        black_state = "";
```

Comment:
I hate to be a DRY zealot, but I think we're talking 150 lines vs 50 lines here.

One approach to reduce duplicate code is to define the coordinate transformations at the outset and loop:

```
tforms = [
    (x, y) => ({x, y}),
    (x, y) => ({x, y: height - y - 1})
        ...
    (x, y) => ({x: y, y: width - x - 1})
    (x, y) => ({x: height - y - 1, y: width - x - 1})
]

ids = []
for (tform of tforms) {
    let black_state = "";
    let white_state = "";
    for (let x = 0; x < width; i++) {
        for (let y = 0; y < height; y++) {
            const c = tform(x, y);
            if (position[row][col] === JGOFNumericPlayerColor.BLACK) {
                black_state += encodeMoves(c.x, c.y)
            }
            if (position[row][col] === JGOFNumericPlayerColor.WHITE) {
                white_state += encodeMoves(c.x, c.y)
            }
        }
    }
    ids.push(`${blck_state}.${white_state}`)
}
// sort and return the ids
```

(one benefit of `encodeMoves` vs `prettyCoords` is it doesn't require a height/width!)

Reply:
Yeah, once again, thanks ... the version submitted is the "make it really easy to understand" version: I consciously decided to dupe rather than loop for that purpose, but your version is elegant.

Reply:
Is there some reason not to just keep the lowest id, and update as we go?

Reply:
Not really, no.  I think I wrote it that way because I actually thought about writing this with a more functional loop:

```
const ids = tforms.map((tform) => {
    /* calculate id for this transform */
    return id;
}
```

(I prefer `map`, but thought maybe for loops are more familiar to other folks?)

Reply:
Yeah - I went through a phase of using functional loops all over the place, but found that unless you have a "good reason" (such as a set of operations on a stream) they can be harder to debug - sometimes because the reasoning is a little more effort, and sometimes just because sometime inserting debug statements is a hassle.